File: 23.合并k个升序链表.c
/*
 * @lc app=leetcode.cn id=23 lang=c
 *
 * [23] 合并K个升序链表
 * 方法一：暴力合并，直接循环，两两合并即可。
 */

// @lc code=start
#include <stdio.h>
#include <stdlib.h>
struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode *mergeTwoLists(struct ListNode *l1, struct ListNode *l2)
{
    struct ListNode *newHead = malloc(sizeof(struct ListNode));
    newHead->next = NULL;
    struct ListNode *cur = newHead;
    struct ListNode *cur1 = l1;
    struct ListNode *cur2 = l2;

    while (cur1 != NULL && cur2 != NULL)
    {
        if (cur1->val < cur2->val)
        {
            cur->next = cur1;
            cur1 = cur1->next;
            cur = cur->next;
        }
        else
        {
            cur->next = cur2;
            cur2 = cur2->next;
            cur = cur->next;
        }
    }
    while (cur1 != NULL)
    {
        cur->next = cur1;
        cur1 = cur1->next;
        cur = cur->next;
    }

    while (cur2 != NULL)
    {
        cur->next = cur2;
        cur2 = cur2->next;
        cur = cur->next;
    }
    return newHead->next;
}

struct ListNode *mergeKLists(struct ListNode **lists, int listsSize)
{
    if (listsSize == 0)
    {
        return NULL;
    }
    if (listsSize == 1)
    {
        return lists[0];
    }
    struct ListNode *result = lists[0];
    for (int i = 1; i < listsSize; i++)
    {
        result = mergeTwoLists(result, lists[i]);
    }
    return result;
}

// @lc code=end

File: LinkedList/23.合并k个升序链表.java
import java.util.ArrayList;
import java.util.Comparator;

/*
 * @lc app=leetcode.cn id=23 lang=java
 *
 * [23] 合并K个升序链表
 * 方法二：将所有结点排入arraylist，然后排序所有结点，重新生成链表。
 */

// @lc code=start

//   Definition for singly-linked list.
public class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists == null || lists.length == 0) {
            return null;
        }
        if (lists.length == 1) {
            return lists[0];
        }
        Comparator<ListNode> comparator = (o1, o2) -> o1.val - o2.val;
        ArrayList<ListNode> arrayList = new ArrayList<ListNode>();
        for (int i = 0; i < lists.length; i++) {
            ListNode head = lists[i];
            while (head != null) {
                arrayList.add(head);
                head = head.next;
            }
        }

        arrayList.sort(comparator);
        for (int i = 0; i < arrayList.size() - 1; i++) {
            arrayList.get(i).next = arrayList.get(i + 1);
        }
        if (arrayList.size() == 0) {
            return null;
        }
        return arrayList.get(0);

    }
}
// @lc code=end
